subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -91,6 +171,799 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
 	}
 }
 
+#if defined(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CONNECTIVITY_CHECK)
+/* Enabling this can cause lot of prints so do it if really needed */
+static bool debug_rx_src;
+
+void conn_mgr_online_checker_update(struct net_pkt *pkt,
+				    void *pkt_hdr,
+				    sa_family_t family,
+				    bool is_loopback)
+{
+	if (!IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_VERIFY)) {
+		return;
+	}
+
+	/* We cannot verify online connectivity from loopback interface */
+	if (is_loopback) {
+		return;
+	}
+
+	if (iface_states[conn_mgr_get_index_for_if(net_pkt_iface(pkt))] == 0) {
+		/* We are not interested in this interface */
+		return;
+	}
+
+	/* Check if the source address of the packet is not local */
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		struct net_ipv6_hdr *hdr = pkt_hdr;
+
+		if (net_ipv6_is_ll_addr((struct in6_addr *)hdr->src)) {
+			return;
+		}
+
+		if (PRIVATE_ADDR_CHECK) {
+			if (net_ipv6_is_ula_addr((struct in6_addr *)hdr->src)) {
+				return;
+			}
+
+			if (net_ipv6_is_private_addr((struct in6_addr *)hdr->src)) {
+				return;
+			}
+		}
+
+		if (debug_rx_src) {
+			NET_DBG("Recv src addr %s", net_sprint_ipv6_addr(hdr->src));
+		}
+
+		online_check->pkts_received = true;
+		net_trickle_consistency(&online_check->verifier);
+
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		struct net_ipv4_hdr *hdr = pkt_hdr;
+
+		if (net_ipv4_is_ll_addr((struct in_addr *)hdr->src)) {
+			return;
+		}
+
+		if (PRIVATE_ADDR_CHECK) {
+			if (net_ipv4_is_private_addr((struct in_addr *)hdr->src)) {
+				return;
+			}
+		}
+
+		if (debug_rx_src) {
+			NET_DBG("Recv src addr %s", net_sprint_ipv4_addr(hdr->src));
+		}
+
+		online_check->pkts_received = true;
+		net_trickle_consistency(&online_check->verifier);
+
+		return;
+	}
+}
+#endif
+
+static void refresh_online_connectivity_check(struct online_check_data *online_data)
+{
+	if (!IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_CONNECTIVITY_CHECK)) {
+		return;
+	}
+
+	/* All interfaces are down. Disable Trickle timer and
+	 * generate offline event.
+	 */
+	net_mgmt_event_notify(NET_EVENT_CONNECTIVITY_OFFLINE, NULL);
+
+	stop_online_check();
+
+	trigger_online_connectivity_check();
+}
+
+static void verifier_cb(struct net_trickle *trickle, bool do_suppress,
+			void *user_data)
+{
+	struct online_check_data *online_data = user_data;
+
+	NET_DBG("Verifier %p callback called (suppress %d)", trickle,
+		do_suppress);
+
+	/* Check if we have received any data between calls to this cb. */
+	if (do_suppress) {
+		NET_DBG("No data transfer seen, not considered online.");
```

Comment:
If I understand correctly, right now in case there's no communication, we'll set the state back to offline and then trigger the connectivity probing, which could report the device is online again if the probe succeeds.

Wouldn't it make sense to do this the other way around? I. e. in case there were no communication from the app activity, we trigger connectivity probing first, and then let the respective connectivity handlers (icmp/http) set the status to offline if they fail?

Reply:
Yes, I suppose there are multiple ways to do this. We have the Trickle library in networking that can be used for triggering connectivity check if there is no activity seen. The criteria for what this "seen" means could be debated, it could be that the application is totally separate and only communicates in certain networks etc. in which case its activity would not count as "online". This is very much debatable and depends on use cases etc.